eosio.unregd/eosio.unregd.cpp

```diff
@@ -21,6 +21,140 @@ void unregd::add(const ethereum_address& ethereum_address, const asset& balance)
   });
 }
 
+/**
+ * Sets the maximum amount of EOS this contract is willing to pay when creating a new account
+ */
+void unregd::setmaxeos(const asset& maxeos) {
+  require_auth(_self);
+
+  auto symbol = maxeos.symbol;
+  eosio_assert(symbol.is_valid() && symbol == CORE_SYMBOL, "maxeos invalid symbol");
+
+  auto itr = settings.find(1);
+  if (itr == settings.end()) {
+    settings.emplace(_self, [&](auto& s) {
+      s.id = 1;
+      s.max_eos_for_8k_of_ram = maxeos;
+    });
+  } else {
+    settings.modify(itr, 0, [&](auto& s) {
+      s.max_eos_for_8k_of_ram = maxeos;
+    });
+  }
+}
+
+/**
+ * Register an EOS account using the stored information (address/balance) verifying an ETH signature
+ */
+void unregd::regaccount(const bytes& signature, const string& account, const string& eos_pubkey_str) {
```

Comment:
Here is the alternative approach to avoid using string operations and parsing the public key string:

1. Generate the `msghash` using the SHA256 hash of the raw packed data of the tuple `(tapos_block_num(), tapos_block_prefix(), new_pubkey, account_name)` where the `account_name` (of `uint64_t` type) would be the second argument of the `regaccount` action and `new_pubkey` (of `eosio::public_key` type) would be the third argument of the `regaccount` action. 
2. Convert the first 30 bytes of `msghash` data into a Base64 encoded string (which should consist of exactly 40 ASCII characters).
3. `memcpy` into the beginning of a `char message[68]` the first 28 bytes of the string `"\x19Ethereum Signed Message:\n40"` and then `memcpy` into `message + 28` the first 40 bytes of the Base64 encoded string generated in step 2.
4. Then calculate the SHA3 hash of message as before: `rhash_keccak_update(&shactx, (const uint8_t*)message, sizeof(message));`.

Steps 1 and 2 would have to be replicated in a client-side tool so that users can simply input their newly desired account name and the new public key (in convenient text format) into the tool which would generate the 40 character Base64-encoded string. The tool would then pause (keeping track of the state of inputs and the TaPoS information it selected) waiting for the signature as input. The 40 character Base64-encoded string would then be signed by the user using their appropriate Ethereum private key and any compatible Ethereum message signing tool before then inputting the outputted signature back into tool to finalize the process by signing and broadcasting the transaction containing the `regaccount` action.

The downside with this approach is that the tool could substitute its own public key (say that of an attacker who tricked users into using their malicious client-side tool) as the input into step 1 and generate a Base64-encoded string that the user would have no reason to suspect has been tampered with. This risk can be mitigated by reminding users to only use client-side tools that have been audited (since the source is open) by multiple people they trust. But the good thing about the current code's approach over this alternative approach is that the user can see the public key and account name that they are putting into the Ethereum message signing tool, so they are less likely to get tricked into giving up their funds to an attacker. The disadvantage of the current code's approach is that it bloats the contract.